### DISCUSSION TITLE
Can I publish a component/library while avoiding duplicate classes?

### DISCUSSION THREAD

**Discussion by richardtallent** [tailwindlabs/tailwindcss#9927] [Help]
I'd like to convert my open-source Vue components to use TW classes. Same with some components I manage at my job (not all of our teams use TW).

One thing that has stopped me from trying is I don't want the final bundled CSS of applications using these components have a bunch of duplicated TW classes due to the application using the component's compiled CSS and *also* using TW for the app itself.

Duplicated classes increase bundle size unnecessarily, but they can also cause some weird issues, since occasionally, declaration order of CSS classes matters (there are some "overlapping" TW classes that rely on this).

So, two questions on this:

1. Is there a way to tell Tailwind's postcss functionality to **remove** any *recognized* TW classes coming in from a component's stylesheet (i.e., a stylesheet in the `node_modules` folder) and **include** them in the application's final TW stylesheet, as if it found those class names in the app's own code? (If the component is using custom variants or colors in its own TW config, obviously TW shouldn't remove and replace those, I'm talking more about the "stock" CSS class names that TW would recognize OOTB.)

2. I don't want to include the TW CSS reset in my component, since doing that at a component level could cause issues with the application. Most sane devs already reset their CSS (whether via TW or otherwise), so at best, it's duplication of effort. Can this be done reasonably in a way that works for both TW and non-TW users of my components?

**Comment by wongjn** [comment]
For consumers that use Tailwind, they could choose not to import the components' CSS at all, and instead scan the components' source code by adding the necessary `node_module` path globs to the Tailwind configuration. This would avoid duplication and rule order issues.

To avoid compiling Tailwind's reset, you could look at [disabling the core preflight Tailwind plugin](https://tailwindcss.com/docs/preflight#disabling-preflight).

**Reply by richardtallent** [reply]
Yes, that works if ALL of the CSS in the component is TW-generated. Doesn't work when there are aspects of the CSS not covered by TW. Also a bit of a hassle for the TW user, having to locate and include all potential class sources.

I was hoping there was a switch I was missing somewhere that would automate this a bit. I'd love to see more component libraries adopt TW, both to reduce unnecessary imported CSS bloat and to encourage consistency in metrics, colors, etc. the easier we can make that, the better! :)
